### PR TITLE
Support running pipeline parallel inference by vertically partitioning model to different devices

### PR DESCRIPTION
**Pull request by yangw1234**
## Description

Support running pipeline parallel inference by vertically partitioning model to different devices. (**Need to install patched IPEX from source**)

1. add to example using `accelerate.dispatch` to partition model into different devices
2. change `LlamaModel.forward` to fix latency problem

### CODE REVIEW

File: python/llm/example/GPU/Pipeline-Parallel-Inference/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+conda install -c conda-forge -y gperftools=2.10 # to enable tcmalloc
+```
+
+### 1.2 Build and install patched version of Intel Extension for PyTorch (IPEX)
+
+```bash
+conda activate llm
+source /opt/intel/oneapi/setvars.sh
+git clone https://github.com/yangw1234/intel-extension-for-pytorch.git
```
Comment (yangw1234) [inline python/llm/example/GPU/Pipeline-Parallel-Inference/README.md:29]: https://github.com/intel/intel-extension-for-pytorch/tree/xpu-main may also work.

File: python/llm/example/GPU/Pipeline-Parallel-Inference/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+### 1.2 Build and install patched version of Intel Extension for PyTorch (IPEX)
+
+```bash
+conda activate llm
+source /opt/intel/oneapi/setvars.sh
+git clone https://github.com/intel/intel-extension-for-pytorch.git
+cd intel-extension-for-pytorch
+git checkout v2.1.10+xpu
+git submodule update --init --recursive
+git cherry-pick be8ea24078d8a271e53d2946ac533383f7a2aa78
```
Comment (yangw1234) [inline python/llm/example/GPU/Pipeline-Parallel-Inference/README.md:33]: @jason-dai changed to official IPEX.

File: python/llm/example/GPU/Pipeline-Parallel-Inference/generate.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    model_path = args.repo_id_or_model_path
+
+    # Load model in 4 bit,
+    # which convert the relevant layers in the model into INT4 format
+    model = AutoModelForCausalLM.from_pretrained(model_path,
+                                                 load_in_4bit=True,
+                                                 optimize_model=True,
+                                                 trust_remote_code=True,
+                                                 use_cache=True)
+    first_half = ['model.embed_tokens', 'model.layers.0', 'model.layers.1', 'model.layers.2',
```
Comment (yangw1234) [inline python/llm/example/GPU/Pipeline-Parallel-Inference/generate.py:65]: This is a little awkward right now. Not sure how to automatically partition the model.

**Comment by jason-dai** [comment]
LGTM. We may need to support Mixtral too.

**Pull request merged**